Skip R packages for Jupyter-only Quarto content

toph-allen opened the pull request:
## Intent

In https://github.com/rstudio/rsconnect/pull/572, Jupyter-only Quarto content incorrectly receives R package dependencies. This PR addresses that issue.

## Approach

The R package dependencies stem from `rsconnect` thinking that, because the Package contains Python chunks in an `.[R|q]md` file, it must require `reticulate`.

In reality, Quarto determines the execution engine on a file-by-file basis. By default, if a document has only Python chunks, it'll use `jupyter`, and if it has R alone or R + Python, it'll use `reticulate`. Users can also force a specific engine in YAML front matter.

`rsconnect` checks for Python Rmds in a small function, `appHasPythonRmd`. We only have Quarto engine info at the project level, and I wanted to avoid adding complexity by attempting to duplicate Quarto's engine binding rules in one conditional fork of that function. Instead, the code used a simpler heuristic: `appHasPythonRmd` will return `FALSE` if `quartoInfo$engines` includes `jupyter` and does not include `knitr`.

- This solves the problem of adding R `packages` in `jupyter`-only Quarto projects.
- This does not solve the problem of adding a `reticulate` dependency in `knitr` + `jupyter` Quarto projects where `knitr` does not execute Python code. However, this is probably somewhat of an edge case, should not cause major issues with environment restoration, and solving it would require either a `quarto` dependency of some sort or really brittle code trying to re-implement Quarto behavior.

This changed meant that I had to move where `pyInfo` is gathered and added to the `manifest` object in `createAppManifest`. Previously, it occurred when looping over the `packrat` snapshot, if the name of the current package was `reticulate`. This meant that `jupyter`-only Quarto projects didn't get `python` information. Now, it occurs after the loop, if `reticulate` is among the packages or if `jupyter` is among Quarto engines.

## Automated Tests

- The final tests in `test-bundle.R` were modified to cover this functionality. 
 
## QA Notes

The code changes in this bundle are localized within `bundle.R`, but affect `writeManifest()` and `deployApp()`. however, the changes in those workflows are identical changes to duplicated code. I *think* it's reasonable to focus on testing `writeManifest()`.

The only changes this PR should see from the initial PR with `writeManifest()` Quarto support  (https://github.com/rstudio/rsconnect/pull/572) is that manifests for Python-only Quarto content should not contain a `packages` object. This in turn should mean that deploying that content to Connect should not require an R environment restore.

No changes should occur in other Quarto content (with the `knitr` engine). No changes should occur to other `reticulated` Python content.

### Testing `writeManifest` changes

You can use the method I described in the prior PR (https://github.com/rstudio/rsconnect/pull/572) to diff changes between versions. You can work with content from the `tests/testthat` directory of this repo, but because you'll want to compare between different branches of this repo, it's easier to copy it to a new directory. I compressed what I think are the relevant subset of those items here: [qa-content.zip](https://github.com/rstudio/rsconnect/files/8383021/qa-content.zip)

1. Extract those items to a new directory. In a terminal, `cd` to that directory and run `git init`. I'll refer to this directory as `qa-content`.
2. Install the version of `rsconnect` from the previous PR (https://github.com/rstudio/rsconnect/pull/572). To do so, open the project in the RStudio IDE, and click the "Install" button in the "Build" tab of the top right pane.
3. In the terminal in `qa-content`, open a new R session. Run the following to write manifests:

```r
# If quarto::quarto_path() doesn't find Quarto, you need to use the IDE quarto
QUARTO_PATH <- "/Applications/RStudio.app/Contents/MacOS/quarto/bin/quarto"
PYTHON <- Sys.which("python")

rsconnect::writeManifest("quarto-doc-none", quarto = QUARTO_PATH)
rsconnect::writeManifest("quarto-proj-r-shiny", quarto = QUARTO_PATH)
rsconnect::writeManifest("quarto-website-py", quarto = QUARTO_PATH, python = PYTHON)
rsconnect::writeManifest("quarto-website-r", quarto = QUARTO_PATH)
rsconnect::writeManifest("quarto-website-r-py", quarto = QUARTO_PATH, python = PYTHON)
rsconnect::writeManifest("shiny-rmds")
rsconnect::writeManifest("shinyapp-simple")
# You'll need RETICULATE_PYTHON configured for this one.
rsconnect::writeManifest("test-reticulate-rmds", , python = PYTHON)
rsconnect::writeManifest("test-rmd-bad-case")
rsconnect::writeManifest("test-rmds")
```

Commit the manifests in this directory.

4. Install _this_ build of `rsconnect. Check out this branch of that repo, and repeat step 2.
5. Start a new R session in `qa-content`. Run the same R commands.
6. Use `git diff` to inspect the changes.

#### Expected changes

- The only change you should see is that `quarto-website-py` should lose its unnecessary `packages` object.
- You should be able to successfully deploy content to RStudio Connect by calling `deployApp()` with the same arguments as `rsconnect` is called above.

### Testing `deployApp` changes

In the directory of content you created above, run the following from an R session.

```r
# Assuming you have a server named "localhost" configured in the IDE / in the rsconnect package
server <- "localhost"

rsconnect::deployApp("quarto-website-py", quarto = quarto::quarto_path(), server = server)
```

This content shouldn't see an R environment restore when deployed with this branch.

The following content items should still see an R environment restore.

```r
# If quarto::quarto_path() doesn't find Quarto, you need to use the IDE quarto
QUARTO_PATH <- "/Applications/RStudio.app/Contents/MacOS/quarto/bin/quarto"

# Assuming you have a server named "localhost" configured in the IDE / in the rsconnect package
server <- "localhost"

rsconnect::deployApp("quarto-proj-r-shiny", quarto = QUARTO_PATH, server = server)
rsconnect::deployApp("quarto-website-r", quarto = QUARTO_PATH, server = server)
rsconnect::deployApp("quarto-website-r-py", quarto = QUARTO_PATH, server = server)
rsconnect::deployApp("shiny-rmds", server = server)
rsconnect::deployApp("shinyapp-simple", server = server)
# You'll need RETICULATE_PYTHON configured for this one, server = server.
rsconnect::deployApp("test-reticulate-rmds", python = Sys.getenv("RETICULATE_PYTHON"), server = server)
rsconnect::deployApp("test-rmd-bad-case", server = server)
rsconnect::deployApp("test-rmds", server = server)
```

aronatkins commented on R/bundle.R:
```diff
@@ -845,6 +852,22 @@ createAppManifest <- function(appDir, appMode, contentCategory, hasParameters,
+      pyInfo$package_manager$contents <- NULL
+    }
+    else {
+      errorMessages <- c(errorMessages, paste("Error detecting python for reticulate:", pyInfo$error))
```
Similar to the idea of the `needsR` helper, maybe we need an `needsPython` helper.

```r
needsPython <- func(quartoInfo) {
  if (is.null(quartoInfo)) {
    # No R-based, non-Quarto content uses Python by default.
    # Looking for Python chunks in Rmd needs to happen separately.
    return(FALSE)
  }
  # Python is a direct consequence of the "jupyter" engine; not "knitr" or "markdown".
  return "jupyter" %in% quartoInfo[["engines"]]
}

needsPyInfo <- reticulateImplied || needsPython()
if (needsPyInfo && !is.null(python)) { ... }
```

I'm also extracting some of the complex conditions into variables.

aronatkins replied:
One final note -- you're not handling `is.null(quartoInfo)` here.

toph-allen replied:
I agree with the overall conditional flow here. I could quibble with some of the names (names are hard!) — like, `needsPython` might be confusing to a reader when what it actually represents here is "uses Jupyter".

I'm not saying that I have a better idea though… I'll implement it and see if anything strikes me.

aronatkins reviewed (commented):
Could you expand the testing notes to explain what types of package dependencies should be seen for different types of content?

For example:

* Rmd with Python chunk
* Rmd without Python chunk
* Shiny without reticulate
* Shiny with reticulate
* Quarto with knitr
* Quarto with markdown
* Quarto with jupyter

When will we see a Python section in the manifest? When will we see R packages? Which are changing based on this PR? Do we expect any changes to the "older" (non-Quarto) content types?

toph-allen commented:
In the [new commits](https://github.com/rstudio/rsconnect/pull/575/files/feed331552273a1c65417d0d23fc3e6701c03e34..55a7179558c023e188bf7494f395199d63953b7b), I made some changes to hopefully improve the code's clarity and readability.

- Line-wrapped one really long string with `paste0`.
- I renamed some existing functions and variables where old names were no longer accurate. If functions were used to assign variables, I tried to keep them verb-y. I tried to be as specific as possible, so distinguishing between documents having Python versus the app using Python. I don't think I did a perfect job, but I think it's an improvement.
    - `hasPythonRmd <- appHasPythonRmd()` becomes `documentsHavePython <- detectPythonInDocuments()`
    - `inferDependencies`, which outputs a list of R packages, becomes `inferRPackageDependencies`.
- Added two new functions, `appUsesR(quartoInfo)` and `appUsesPython(quartoInfo)`. I considered using `content`, but decided to follow the rest of the package in referring to the content as `app`. I considered using `needsR()` and `needsPython()`, but since R content that uses `reticulate` could also be said to "need Python", I tried to get more specific with my naming. I considered `engineNeedsR` and `engineNeedsPython`, but "engine" is a Quarto-specific concept and these functions might apply to other content types in the future. So, `app` distinguishes it from `document` in `documentsHavePython`, and `Uses` is trying to suggest that this is the primary thing the content runs, rather requiring it as a downstream dependency.
- It wound up not making sense to make an `impliesReticulate()` function or variable, since reticulate is detected in a conditional in `inferRPackageDependencies()`, and I wanted to keep the results of that function as the only place where we try to infer that dependency. Instead, I wrote `needsPyInfo <- appUsesPython(quartoInfo) || "reticulate" %in% names(packages)`.

I don't feel like these changes drastically improve fundamental problems, but they at least update some of the names and present further fragmentation of functionality.

toph-allen commented:
The most recent changes fix a bug that @ChaitaC and I saw during validation, where R packages were not omitted for `quarto-website-py`. It turns out that in her environment, `snapshotDependencies()` (which is now named `snapshotRDependencies()`) was adding dependencies for `BiocManager` greedily, even if that package was not loaded.

I gated the snapshot function behind `appUsesR(quartoInfo)`, which right now only returns false if Quarto info is present and `knitr` is not in the list of engines. This change should be pretty tightly scoped, but I think it'd be good to get @aronatkins's take on it.

@ChaitaC QA should be ready to go on this again.

aronatkins reviewed (approved):
Would you mind sharing where the bioc dependencies were injected?

toph-allen commented:
@ChaitaC https://github.com/rstudio/rsconnect/pull/576 also contains the changes from this branch.

toph-allen commented:
@aronatkins  I dug deeper into the code. It looks like the vendored `renv` dependency discovery code will assume a `quarto` dependency if it finds Quarto files in the directory.

https://github.com/rstudio/packrat/blob/e0bd2ec38ac393fdff96c3cf063b11bad6bf02af/inst/resources/renv.R#L5012-L5029

This explains why @ChaitaC was seeing the packages appear and I wasn't — I was using CRAN `packrat`.
